framework/src/modules/chain/synchronizer/synchronizer.js

```diff
@@ -14,47 +14,175 @@
 
 'use strict';
 
-const BlockSynchronizationMechanism = require('./block_synchronization_mechanism');
-const FastChainSwitchingMechanism = require('./fast_chain_switching_mechanism');
+const util = require('util');
+const assert = require('assert');
+
+const {
+	verifySignature,
+	verifyVersion,
+	verifyReward,
+	verifyId,
+	verifyPayload,
+} = require('../blocks/verify');
 
 class Synchronizer {
-	constructor() {
-		this.currentStrategy = null;
-		this.blockSynchronizationMechanism = new BlockSynchronizationMechanism();
-		this.fastChainSwitchingMechanism = new FastChainSwitchingMechanism();
+	constructor({
+		storage,
+		logger,
+		blockReward,
+		exceptions,
+		maxTransactionsPerBlock,
+		maxPayloadLength,
+	}) {
+		this.storage = storage;
+		this.logger = logger;
+		this.blockReward = blockReward;
+		this.exceptions = exceptions;
+
+		this.constants = {
+			maxPayloadLength,
+			maxTransactionsPerBlock,
+		};
+
+		this.mechanisms = [];
+	}
+
+	/**
+	 * Register a sync mechanism with synchronizer
+	 * It must have "isValidFor" and "run" interface to call upon
+	 *
+	 * "isValidFor" must return true/false to match the sync mechanism
+	 * "isValidFor" can throw error to abort the synchronization
+	 *
+	 * "run" must initiate the synchronization
+	 * "run" must keep track of its state internally
+	 *
+	 * @param {Object} mechanism - Mechanism to register
+	 */
+	register(mechanism) {
+		assert(
+			util.types.isAsyncFunction(mechanism.isValidFor),
+			'Sync mechanism must have "isValidFor" async interface'
+		);
+		assert(
+			util.types.isAsyncFunction(mechanism.run),
+			'Sync mechanism must have "run" async interface'
+		);
+
+		// Check the property isActive, it can be own property or a getter
+		assert(
+			mechanism.isActive !== undefined,
+			'Sync mechanism must have "isActive" interface'
+		);
+
+		this.mechanisms.push(mechanism);
 	}
 
 	/**
 	 * Start the syncing mechanism
+	 * @param {Object} receivedBlock - The block you received from network, used to choose sync mechanism
 	 * @return {*}
 	 */
-	run() {
-		if (this.currentStrategy && this.currentStrategy.isActive()) {
-			throw new Error('Blocks Sychronizer is already running');
+	async run(receivedBlock) {
+		if (this.activeMechanism) {
+			throw new Error(
+				`Blocks Sychronizer with ${
+					this.activeMechanism.constructor.name
+				} is already running`
+			);
+		}
+
+		const lastBlock = await this.storage.entities.Block.getLastBlock();
+
+		// Moving to a Different Chain
+		// 1. Step: Validate new tip of chain
+		const result = this._verifyBlockBeforeSync(lastBlock, receivedBlock);
```

Comment:
The checks included in this step are validation steps according to the definition of validate and verify, so i think we should use validate here.

Reply:
If you see functions inside `verifySignature`, `verifyVersion`, `verifyReward`, `verifyId`, `verifyPayload`, then the name make sense. Otherwise we would have `validate*` for all other functions as well.

Reply:
I agree it's being done incorrectly for block validation because it is an old code. But recently we had strictly defined the distinction b/w validate and verify which can be seen when we are checking transactions. There we group all nondb related checks in the validate step and db related checks in the verify/apply steps.

Reply:
Hence we should follow the new approach

Reply:
Is there any issue created to refactor all code for block verification? If not then we should create it and also make refactor this as well under same issue.

Reply:
I have created an issue for to rename functions for block here https://github.com/LiskHQ/lisk-sdk/issues/4025. Please update the function name here since this is new implementation.

Reply:
@SargeKhan  Done. Please review.